tests/testthat/test-type-dplyr.R

```diff
@@ -0,0 +1,144 @@
+
+# `grouped_df` -------------------------------------------------------
+
+test_that("grouped-df is proxied and restored", {
+  gdf <- dplyr::group_by(mtcars, cyl)
+
+  expect_identical(vec_proxy(gdf), gdf)
+  expect_identical(vec_restore(mtcars, gdf), gdf)
+
+  expect_identical(vec_ptype(gdf), gdf[0, ])
+
+  # Taking the prototype with `[` because of tidyverse/dplyr#5040
+  gdf <- dplyr::group_by(mtcars, cyl, am, vs)
+  # expect_identical(gdf[0, ], vec_ptype(gdf))
+
+  out <- vec_ptype(dplyr::group_by(mtcars, cyl, .drop = FALSE))
+  expect_drop(out, FALSE)
+})
+
+test_that("can take the common type of grouped tibbles and tibbles", {
+  gdf <- dplyr::group_by(mtcars, cyl)
+  expect_identical(vec_ptype2(gdf, data.frame()), vec_ptype(gdf))
+  expect_identical(vec_ptype2(data.frame(), gdf), vec_ptype(gdf))
+  expect_identical(vec_ptype2(gdf, tibble()), vec_ptype(gdf))
+  expect_identical(vec_ptype2(tibble(), gdf), vec_ptype(gdf))
+
+  gdf_nodrop <- dplyr::group_by(mtcars, cyl, .drop = FALSE)
+  expect_drop(vec_ptype2(gdf, gdf_nodrop), FALSE)
+  expect_drop(vec_ptype2(gdf_nodrop, gdf), FALSE)
+  expect_drop(vec_ptype2(gdf_nodrop, mtcars), FALSE)
+  expect_drop(vec_ptype2(mtcars, gdf_nodrop), FALSE)
+})
+
+test_that("the common type of grouped tibbles includes the union of grouping variables", {
+  gdf1 <- dplyr::group_by(mtcars, cyl)
+  gdf2 <- dplyr::group_by(mtcars, am, vs)
+  expect_identical(
+    # Taking the prototype again because of tidyverse/dplyr#5040
+    vec_ptype(vec_ptype2(gdf1, gdf2)),
+    vec_ptype(dplyr::group_by(mtcars, cyl, am, vs))
+  )
+})
+
+test_that("can cast to and from `grouped_df`", {
+  gdf <- dplyr::group_by(unrownames(mtcars), cyl)
+  input <- mtcars[10]
+  cast_gdf <- dplyr::group_by(vec_cast(mtcars[10], mtcars), cyl)
+
+  expect_error(
+    vec_cast(input, dplyr::group_by(mtcars["cyl"], cyl)),
+    class = "vctrs_error_cast_lossy"
+  )
+
+  expect_identical(
+    vec_cast(input, gdf),
+    cast_gdf
+  )
+  expect_identical(
+    vec_cast(gdf, mtcars),
+    unrownames(mtcars)
+  )
+
+  expect_identical(
+    vec_cast(tibble::as_tibble(input), gdf),
+    unrownames(cast_gdf)
+  )
+  tib <- tibble::as_tibble(mtcars)
+  expect_identical(
+    unrownames(vec_cast(gdf, tib)),
+    tib
+  )
+})
+
+test_that("casting to `grouped_df` doesn't require grouping variables", {
+  expect_identical(
+    vec_cast(mtcars[10], dplyr::group_by(mtcars, cyl)),
+    dplyr::group_by(vec_cast(mtcars[10], mtcars), cyl)
+  )
+})
+
+test_that("casting to `grouped_df` handles `drop`", {
+  gdf_nodrop <- dplyr::group_by(mtcars, cyl, .drop = FALSE)
+  expect_identical(vec_cast(mtcars, gdf_nodrop), gdf_nodrop)
+})
+
+test_that("can cbind grouped data frames", {
+  gdf <- dplyr::group_by(mtcars[-10], cyl)
+  df <- unrownames(mtcars)[10]
+
+  expect_identical(
+    unrownames(vec_cbind(gdf, df)),
+    tibble::as_tibble(mtcars)[c(1:9, 11, 10)]
+  )
+})
```

Comment:
Also, it is somewhat interesting that `vec_ptype2(rww, gdf)` is an error, but `vec_cbind(rww, gdf)` works and returns a rowwise_df because the `df_container_type()` of `rww` is a rowwise_df, and for `gdf` it is a simple tibble.

This somehow feels wrong?

Reply:
Good point. I agree it doesn't feel right, but I'm not sure how to do better. The alternative is to decide the frame ptype of rowwise data frames is a tibble purely for consistency with grouped data frames. This feels arbitrary as well.

Reply:
My feeling is that we can't do better without a proper frame API. Unfortunately there's too much work left and uncertainty around it for the dplyr 1.0 release.